gh-142466: Pixi package definitions for downstream use (#142469)
python/cpython: Tools/pixi-packages/README.md

> - see comments for which fields should be maintained and kept up-to-date

Reviewer: When I read this, I still have no idea what the stuff in this folder is doing/for, can you please add a general overview. Also, you need to add an entry to the top-level `Tools/README`

Author: yeah of course, just opened this for preliminary discussion first.

Author: @StanFromIreland please could you review https://github.com/python/cpython/pull/142469/commits/b49095c1bc6ffdd7246e5aef4b99ab9e6411cf3b ? Thanks!

> - combining directories blocked on https://github.com/prefix-dev/pixi/issues/4599

Author: hopefully once this issue is solved, we would be able to have just one `pixi.toml` and `recipe.yaml` file, rather than separate ones in separate directories for each build variant.

> ## CPython Pixi Packages

Reviewer 2: ```suggestion
# CPython Pixi packages
```

This directory contains definitions for [Pixi packages](https://pixi.sh/latest/reference/pixi_manifest/#the-package-section)
which can be built from the CPython source code.

> Downstream developers can make use of these packages by adding them as git dependencies in a

Reviewer 2: ```suggestion
Downstream developers can make use of these packages by adding them as Git dependencies in a
```

[Pixi workspace](https://pixi.sh/latest/first_workspace/), like:

```toml
[dependencies]
python = { git = "https://github.com/python/cpython", subdirectory = "Tools/pixi-packages/asan" }
```

This is particularly useful when developers need to build CPython from source
(for example, for an ASan-instrumented build), as it does not require any manual

> cloning or building steps. Instead, Pixi will automatically handle both the building

Reviewer 2: ```suggestion
clone or build steps. Instead, Pixi will automatically handle both the build
```

and installation of the package.

Each package definition is contained in a subdirectory, but they share the build script
`build.sh` in this directory. Currently defined package variants:

- `default`
- `asan`: ASan-instrumented build with `PYTHON_ASAN=1`

> ### Maintenance

Reviewer 2: ```suggestion
## Maintenance
```

- the `version` fields in each `recipe.yaml` should be kept up to date with the Python version
- dependency requirements should be kept up to date in each `recipe.yaml`

> - `build.sh` should be updated for any breaking changes in the `configure` and `make` workflow

Reviewer 2: ```suggestion
- Keep the `version` fields in each `recipe.yaml` up to date with the Python version
- Keep the dependency requirements up to date in each `recipe.yaml`
- Update `build.sh` for any breaking changes in the `configure` and `make` workflow
```

> ### Opportunities for future improvement

Reviewer 2: ```suggestion
## Opportunities for future improvement
```

- more package variants (e.g. TSan, UBSan)
- support for Windows
- using a single `pixi.toml` and `recipe.yaml` for all package variants is blocked on https://github.com/prefix-dev/pixi/issues/4599

> - a workaround can be removed from the build script once https://github.com/prefix-dev/rattler-build/issues/2012 is resolved

Reviewer 2: ```suggestion
- More package variants (such as TSan, UBSan)
- Support for Windows
- Using a single `pixi.toml` and `recipe.yaml` for all package variants is blocked on https://github.com/prefix-dev/pixi/issues/4599
- A workaround can be removed from the build script once https://github.com/prefix-dev/rattler-build/issues/2012 is resolved
```
